app/models/version.rb

```diff
@@ -37,18 +37,18 @@ class Version < ApplicationRecord
   ].freeze
 
   validates :project_id, :number, presence: true
-  validates :number, uniqueness: { scope: :project_id }
```

Comment:
this validation was added on [9/4/2015](https://github.com/librariesio/libraries.io/commit/70786598de399c381b15427ec8476d7c446e0716), and we have an index that was [created on 9/5/2015](https://github.com/librariesio/libraries.io/commit/7ce30a0a49ba3c73282ddb146fe4ac19b4d1e254). 

The validation was creating an N+1 since it had to run a `SELECT 1 FROM versions WHERE ...` for each version before inserting, and I don't see a reason why we'd need the validation if we already have the index (considering that `PackageManager::Base.update` is the central place that Versions are created). LMK if my assumptions are wrong here tho

Reply:
Probably we should have a linter rule banning this validation. It is either unnecessary (because of the unique index) or a race condition (because no unique index). Or really it's a race even with the index. Rails should be embarrassed about even having this feature.

Reply:
I think it's because the original idea was to have all your validations/callbacks in ruby-land, e.g. rails didn't even have foreign-key support until [version 4.2](https://github.com/rails/rails/pull/15606) so there are probably a ton of legacy apps that don't have fks. 

I vaguely remember something about the idea that AR would just give you a Hash-like interface (this is pre-NoSQL) and you ideally wouldn't have to worry about the underlying implementation (hah). 

Also possible that it goes back to the original intention of ActiveRecord to be database-agnostic, so some of the database adapters might not have supported unique indices at the time (I'm just speculating tho). 

A rubocop rule for this is a good performance idea tho

Reply:
I think it could just be the opposite of the rules in this cop: https://github.com/rubocop/rubocop-rails/blob/9d5a22383e7d0177a6bee3ec5a095ff590e20ab6/lib/rubocop/cop/rails/unique_validation_without_index.rb#L34-L40